### ISSUE TITLE
[IntelliSense] [Salsa] Support @callback in JSDoc

### ISSUE THREAD

**Issue by egamma** [Microsoft/TypeScript#7515]
_From @santiagohdzb on March 15, 2016 3:43_

Hi everyone, I'm in love with the new VS Code's Salsa, but I'm facing this issue:

![asasdadasd](https://cloud.githubusercontent.com/assets/8533075/13767114/bc11993e-ea24-11e5-8eb2-a6e37ee5ce8d.png)

![asasdadasd](https://cloud.githubusercontent.com/assets/8533075/13767230/b1b0ae0c-ea25-11e5-8b8b-54c139354766.png)

I'm trying to use InselliSense with [JSDoc](http://usejsdoc.org/tags-param.html#callback-functions)'s callback documentation.

Am I doing something wrong? How am I supposed to do this?
- VSCode Version: 0.10.11
- OS Version: Win 10

Steps to Reproduce:
    1. Open a new Javascript file
    2. add the code you see above

_Copied from original issue: Microsoft/vscode#4223_

**Comment by egamma** [comment]
_From @anseki on March 15, 2016 5:6_

Hi @santiagohdzb,
Now, it seems that IntelliSense indicates some types other than primitive data type and `Object` as `any`.
Some types of JavaScript's built-in-object (e.g. `Date`, `RegExp`, etc.) are recognized, but some types (e.g. `Array`, `Map`, etc.) are not.
https://github.com/Microsoft/vscode/issues/3802

**Comment by egamma** [comment]
Moving to TypeScript for investigation.

**Comment by pisrael** [comment]
I'm having the same problem with JSCode. Any workaround?

**Comment by mhegazy** [comment]
We have recently added support for [JSDOC @typedef](https://github.com/Microsoft/TypeScript/pull/8103). if you are using the [nightly build](https://github.com/Microsoft/TypeScript-Handbook/blob/master/pages/Nightly%20Builds.md#visual-studio-code), you can replace the pattern above with:

```ts
/**
 * @typedef {function(number)} addedCallback
 */

/**
 *@param {addedCallback} a
 */
var add = function (a) { 
    
}
```

**Comment by sant123** [comment]
@mhegazy is there a way to do this with TypeScript v1.8.10? Or only for the next release of TypeScript 2.0?

Thanks.

**Comment by zhengbli** [comment]
@sant123 support for `@typedef` is added only in 2.0, and won't be ported back to 1.8.10. You can use the nightly build via `npm install typescript@next -g` to use it today.

**Comment by sant123** [comment]
Thank you @zhengbli :smile:

**Comment by ManvendraSK** [comment]
Any progress on this?

**Comment by episage** [comment]
+1

**Comment by slawo** [comment]
+1
typedef is very limited and doesn't change anything compared to describing the callback inline:

    /**
     *@param {function(int)} a
     */
    var add = function (a) { 
    
    }

The real use of `@callbak` is to properly describe all the parameters of the callback.

**Comment by vsimonian** [comment]
Hate to be that person, but it's been 3 months since the last comment on this issue. Any updates, blocking issues? Is it backlogged, perhaps?

**Comment by mhegazy** [comment]
it is on our Future milestone. so no specific release has been marked. would be happy to take a PR for it anytime though.

**Comment by iugo** [comment]
@mhegazy How to name arg0, arg1 for `function(number, string)`?

**Comment by JobLeonard** [comment]
@iugo: the only workaround right now is to create "dummy" functions, see [this earlier comment](https://github.com/Microsoft/TypeScript/issues/7515#issuecomment-299950195) by @slawo.

It's a bit crude, but as long as you use a bundler like WebPack, which can do dead code elimination upon elimination, you shouldn't have to worry about those dummy functions it ending up in your production code.

Here's more elaborate example, extracted fron [a bigger code snippet](https://gist.github.com/JobLeonard/f3f612fd0ac325992d59b94e55700b7e):

```js
/**
 * A compare function to pass to `indices.sort()`, see also:
 * [`TypedArray.prototype.sort()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/TypedArray/sort).
 *
 * @param {number} i
 * @param {number} j
 */
function indexCompareFunction(i, j) {
  // dummy function for missing VSCode JSDoc callback support
  return i - j;
}

// Explicit
/**
 * Creates a compare function to sort a _separate_ array
 * of indices, based on lexicographical ordering of
 * the array values. Produces a stable sort.
 * @param {*[]} array
 * @returns {indexCompareFunction}
 */
function baseCompareFunc(array) {
  return (i, j) => {
    let vi = array[i],
      vj = array[j];
    return vi < vj ?
      -1 : vi > vj ?
        1 : i - j;
  };
}

// Inferred from callback JSDoc
/**
 * @param {*[]} array
 */
function baseCompareFunc2(array) {
  // Must store the function in a variable,
  // otherwise JSDoc does nothing!
  /**
   * @param {number} i
   * @param {number} j
   */
  let compareFunc = (i, j) => {
    let vi = array[i],
      vj = array[j];
    return vi < vj ?
      -1 : vi > vj ?
        1 : i - j;
  };
  return compareFunc;
}
```

**Comment by mhegazy** [comment]
> @iugo: the only workaround right now is to create "dummy" functions, see this earlier comment by @slawo.

@JobLeonard  this is not accurate. you can do this today using `function`, e.g.:

```js
/**
 * Creates a compare function to sort a _separate_ array
 * of indices, based on lexicographical ordering of
 * the array values. Produces a stable sort.
 * @param {*[]} array
 * @returns {function(number, number): number}
 */
function baseCompareFunc(array) {
    return (i, j) => {
      let vi = array[i],
        vj = array[j];
      return vi < vj ?
        -1 : vi > vj ?
          1 : i - j;
    };
  }
  
```


![image](https://user-images.githubusercontent.com/8000722/31957644-4d0d39d0-b8a4-11e7-8fac-165b9e875532.png)

**Comment by JobLeonard** [comment]
@mhegazy: You're missing the point, the question was:

> How to name `arg0`, `arg1` for `function(number, string)`?

![image](https://user-images.githubusercontent.com/259840/31958027-be09ba86-b8f0-11e7-9565-f4c25160cf5d.png)

I don't see any other way to get named arguments in there at the moment.

**Comment by mhegazy** [comment]
> How to name arg0, arg1 for function(number, string)?

```ts
/**
 * @param {*[]} array
 * @returns {{(i:number, j:number)=> number}}
 */
```
![image](https://user-images.githubusercontent.com/8000722/31958530-ebd08624-b8a6-11e7-895f-cb6318414ba5.png)

**Comment by JobLeonard** [comment]
Wow, nice! That `{{ () => .. }}` notation is new to me though, and skimming through the [JSdoc pages](http://usejsdoc.org/) doesn't give me any hints either - in fact [a sitewide search](https://encrypted.google.com/search?hl=en&q=site%3Ahttp%3A%2F%2Fusejsdoc.org%2F%20{{) gives just one hit, and it's not this.

**Comment by mhegazy** [comment]
inside `{ .. }` we allow TypeScript type syntax, and that is how you define a call signature in TS.

**Comment by JobLeonard** [comment]
Ah, so it's _technically_ not valid JSDoc syntax, and this will only work in VSCode?

I mean, that's perfectly fine by me, but it might be important to some.

**Comment by mhegazy** [comment]
yes. it is not _standard_ JSDoc (thought is very hard to define what is a "standard" in JSDoc). We still have this issue open to support the `@callback` natively.

**Comment by amedley** [comment]
I too could really use the native JSDoc `@callback` syntax because not everyone I work with uses VSC (so we require native documentation). For now I am just setting all callback parameters in documentation to `@param {function} callbackName` so that when this is fixed I can easily go back and find all occurrences of `@param {function}` to tidy up documentation.

**Comment by MichaelMitchellM** [comment]
Hi, this is still a problem in 1.21.0-insiders.
Has any progress been make on this?

**Comment by Gochida** [comment]
> yes. it is not standard JSDoc (thought is very hard to define what is a "standard" in JSDoc). We still have this issue open to support the @callback natively.

At least when I needed to find out how to define a callback in jsdoc, the path to doing it was very well laid out and @callback was the top result for "how to document callbacks in jsdoc". I'm not sure how more 'standard' it needs to be considered when [usejsdoc.org[(http://usejsdoc.org/tags-callback.html) lays out what @callback is and how it works - and is the top result in google searches. I think it'd be reasonable to assume @callback should work the way usejsdoc.org defines it.

@typedef is a poor substitute with no way to define argument names properly. I can either do it the way google is saying others do it and leave people using my code in VSCode high and dry, or incompletely document it the way VSCode allows and leave nothing for those using an IDE with more complete JSDoc support.

Neither seems like a happy solution to me. The purpose of good portable documentation is lost if only part of jsdoc is implemented. This issue has existed for 2 years now, how much longer is it going to take to support SOME kind of callback documentation in VSCode? Callbacks are a major part of JS AND TypeScript - not being able to have documentation half as good as you get in another IDE is silly at this point. Even more silly when VSCode is meant to be the defacto IDE for languages like TypeScript.

VSCode could come out with it's own special @VSCodeCallback if need be - as long as there's something, I could then define the callback twice - vscode's way, and the way everyone else does it - and no one is left with incomplete documentation about my callbacks then.

**Comment by TheLarkInn** [comment]
@DanielRosenwasser @mhegazy this would be pretty nice for us also since we are incredibly callback heavy.

**Comment by sandersn** [comment]
Fix is up at #23947
